src/fraud_handler.py

```diff
@@ -0,0 +1,109 @@
+from botocore.errorfactory import ClientError
+import random
+import string
+
+
+class VerifyFraudToS3:
+
+
+    def __init__(
+        self,
+        key,
+        bucket,
+        role,
+        s_index,
+        s_sourcetype="json",
+        s_source="null",
+        s_host="null"
+        ):
+
+        self.key = key
+        self.bucket = bucket
+        self.role = role
+        self.s_index = s_index
+        self.s_sourcetype = s_sourcetype
+        self.s_source = s_source
+        self.s_host = s_host
+
+
+    def base_json_event(self):
+        return {
+            "host": self.s_host,
+            "source": self.s_source,
+            "sourcetype": self.s_sourcetype,
+            "index": self.s_index,
+            "event": {},
+            }
+
+
+    def aws_session(self, role, session_name='session_verify_fraud_logging'):
+        logger = logging.getLogger('assume-role')
+        logger.setLevel(logging.INFO)
+
+        try:
+            client = boto3.client('sts')
+            response = client.assume_role(RoleArn=self.role, RoleSessionName=session_name)
+            session = boto3.Session(
+                aws_access_key_id=response['Credentials']['AccessKeyId'],
+                aws_secret_access_key=response['Credentials']['SecretAccessKey'],
+                aws_session_token=response['Credentials']['SessionToken'])
+            return session
+        except ClientError as e:
+            logger.error('Unable to assume a session. {0}'.format(e))
+            raise e
+
+    def push_to_s3(self, payload):
+        logger = logging.getLogger('push-to-s3')
+        logger.setLevel(logging.INFO)
+
+        session_assumed = self.aws_session(self.role)
+        logger.info(session_assumed.client('sts').get_caller_identity()['Account'])
+
+        s3 = session_assumed.resource('s3')
+
+        if isinstance(payload, str):
+            payload = json.loads(payload)
+        json_event = self.base_json_event()
+
+        try:
+            if "timestamp" in payload:
+                json_event["time"] = payload["timestamp"]
+        except TypeError as te:
+            logger.info('Timestamp not set. {0}'.format(te))
+            pass
+
+        json_event["event"] = json.dumps(payload)
+        event = str(json.dumps(json_event))
+
+        try:
+            s3.Object(self.bucket, self.key).put(
+                    Body=(bytes(json.dumps(event, indent=2).encode('UTF-8')))
+                )
+        except ClientError as e:
+            logger.error('Unable to write to S3 bucket. {0}'.format(e))
+            raise e
+
+
+def fraud_logging_handler(event, context):
+
+    bucket = os.environ['SPLUNK_VERIFY_FRAUD_S3_NAME']
+    role = os.environ['SPLUNK_S3_BUCKET_ROLE']
+    s_index = os.environ['SPLUNK_INDEX']
+    today = str(datetime.now())
+    letters = string.ascii_lowercase
+    random_string = ''.join(random.choice(letters) for i in range(8))
+    key = 'verify-fraud-events-{}-{}.log'.format(random_string, today)
```

Comment:
It looks like we're generating a new file (with a random name) for every event. Is that definitely what you intended?

Reply:
Right: S3 objects aren't appendable, which is what I thought in the first place. So we have to use a different filename each time.

It might be worth considering using `event_id` instead as that's already guaranteed to be unique (and if we do find ourselves writing the same event twice, it'd be good to see a hard failure and find out!).

For the volumes we're talking about (low) I'm sure this is fine, I guess I just never considered using S3 to write individual log lines as separate S3 objects, and it still feels a bit strange to me.

Reply:
yes this is correct, its been a while since this branch was put together we couldn't remember how each event was placed in s3, but correct a new object is created for each event, we can make the change to have each key named as the unique event id, good shout